# Reproducibility Workflow :two_men_holding_hands: :chart_with_upwards_trend:

Reproducibility in bioinformatics research is based on three facets;

   a) **Designing and planning for the research**
   - This involves organizing relevant literature using tools such as Mendeley, create project management environment using platforms such as the Open Science Network 
   and writing a data management plan.
   - Notes on the experiment an also be stored in an electronic lab notebook.
   
   b)  **Collecting and Analyzing Data**
   - This involves documentationn (use of literate programming tools such as Jupyter Notebooks and R Markdown) and version control (using tools such as Git and GitHub).
   - Data like sequencing data can be clearly structured in spreadsheets with well indicated rows and columns.
   - Workflows can be created using platforms such as SnakeMake and Galaxy and containerized bioinformatics can also be used which allows the researcher to have a precisely controlled environment with all the resources he/she requires. (Examples include Docker and Conda)
   
   c) **Publishing and sharing your work**
   - This uses specific software that allows for collaborative efforts and open access asfter publishing.
   - Data can also be made available to public repositories such as the NCBI Sequence Read Archive (SRA) and the EMBL-EBI European Nucleotide Archive (ENA) which are mostly used to store raw sequence data from NGS projects and alignment information in form of read placements which are readily available to the public.
   
